# **Limpeza para análise e treino de modelo -  Base Lending Club**

## **Aluno:** Rafhael de Oliveira Martins

### **Ferramentas Utilizadas**
* Google Colab
* Google Drive

### **Base de Dados**
* 2Gb

# 1 - Entendimento do Negócio
## Contexto
Você foi procurado por um cliente que precisa dos seus serviços como Cientista de Dados. Ele possui algum dinheiro para investir e achou uma boa ideia alocar parte do dinheiro disponível em crowd funding e, para isso, optou por realizar aportes em pedidos de financiamento no site Lending Club.

Esse site, Lending Club, funciona como um portal onde pessoas que precisam de empréstimos realizam um extenso cadastro indicando, entre outras coisas, o quanto precisa, quanto gostaria de pagar de juros e em quantas vezes pretende pagar. Feito esse cadastro o pedido de empréstimo fica visível para os investidores que podem aportar dinheiro no pedido de empréstimo, recebendo parte dos juros pagos pelo tomador do empréstimo como lucros.


## Objetivo
Sabendo que trata-se de um modelo de classificação binária e que podem ser aplicados diversos modelos estatísticos para geração desse tipo de análise, faça a limpeza dos dados.
* Você deve determinar quais campos serão utilizados para o modelo e quais outros campos devem ser descartados (as discussões levantadas em sala de aula podem ser levadas em consideração para determinar quais dados serão mantidos ou descartados);
* Os campos devem ser convertidos em inteiro ou número real;
* Podem ser gerados dummies ao seu próprio critério;
* Podem ser gerados tokens ao seu próprio critério;
* Campos do tipo “object” devem ser evitados a todo custo;
* O resultado da limpeza deve ser salvo no formato pikle.

# 2 - Entendimento dos Dados
## 2.1 - Carregando a base de dados

Por estar em ambiente de nuvem e os arquivos serem grandes em espaço de armazenamento para upload, optei por utilizar a leitura da base de dados através do Google Drive com a biblioteca from google.colab import drive com o parâmetro drive.mount('/content/drive') em que este monta uma conexão com o Google Drive da conta a ser escolhida e fornece acesso aos arquivos desta conta que foi fornecida autorização.

* Documentação: [Google Drive para Google Colab](https://colab.research.google.com/notebooks/io.ipynb)
* Procurar por: Google Drive -> Montar Google Drive localmente.

In [1]:
# Abrindo conexão com o Google Drive.

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
# Importando Bibliotecas
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pickle

#Apresentas os dados no notebook.
%matplotlib inline 

# Tamanho da imagem.
plt.rcParams['figure.figsize'] = (12,8)

# Parametros do Pandas, limitando a quantidade máxima e a largura das colunas.
pd.set_option('max_columns', 120)
pd.set_option('max_colwidth', 5000)

### 2.1.1 - Lendo o dicionário de dados

In [ ]:
# Lendo o dicionário de dados.
dicionario = pd.read_csv('/content/drive/MyDrive/ArquivosPA/limpeza_analise_modelo_Lending_Club/Cópia de LCDataDictionary_Traduzido.csv', sep=';')

In [ ]:
# Informacoes do dicionario.
dicionario.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151 entries, 0 to 150
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   LoanStatNew  151 non-null    object
 1   Descrição    151 non-null    object
dtypes: object(2)
memory usage: 2.5+ KB


In [ ]:
# Primeiros dados do dicionario.
dicionario.head()

LoanStatNew  \
0        acc_now_delinq   
1  acc_open_past_24mths   
2            addr_state   
3              all_util   
4            annual_inc   

                                                                 Descrição  
0                   O número de contas em que o devedor está inadimplente.  
1                       Número de transações abertas nos últimos 24 meses.  
2                 O estado fornecido pelo mutuário no pedido de empréstimo  
3                     Saldo para limite de crédito em todas as negociações  
4  A renda anual auto-relatada fornecida pelo mutuário durante o registro.

In [ ]:
# Ultimos dados do dicionario.
dicionario.tail()

LoanStatNew  \
146      settlement_status   
147        settlement_date   
148      settlement_amount   
149  settlement_percentage   
150        settlement_term   

                                                                                                                    Descrição  
146  O status do plano de liquidação do mutuário. Os valores possíveis são: COMPLETE, ACTIVE, BROKEN, CANCELED, DENIED, DRAFT  
147                                                               A data em que o mutuário concorda com o plano de liquidação  
148                                                                O valor do empréstimo que o mutuário concordou em liquidar  
149                                   O valor da liquidação como uma porcentagem do valor do saldo de reembolso do empréstimo  
150                                                            O número de meses que o mutuário estará no plano de liquidação

### 2.1.2 - Lendo o dataset da Lending Club

In [ ]:
# Lendo os dados da Leading Club.
dados = pd.read_csv('/content/drive/MyDrive/ArquivosPA/limpeza_analise_modelo_Lending_Club/Cópia de Loan_status_2007-2020Q3.csv', engine='c')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (1,48,58,117,127,128,129,132,133,134,137) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [ ]:
# Informacoes dos dados.
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925493 entries, 0 to 2925492
Columns: 142 entries, Unnamed: 0 to debt_settlement_flag
dtypes: float64(106), int64(1), object(35)
memory usage: 3.1+ GB


In [ ]:
# Proporção dos dados em linhas e colunas
dados.shape

(2925493, 142)

In [ ]:
# Primeiros dados.
dados.head()

Unnamed: 0       id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0           0  1077501     5000.0       5000.0           4975.0   36 months   
1           1  1077430     2500.0       2500.0           2500.0   60 months   
2           2  1077175     2400.0       2400.0           2400.0   36 months   
3           3  1076863    10000.0      10000.0          10000.0   36 months   
4           4  1075358     3000.0       3000.0           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-2011   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-2011  Charged Off   
2           RENT     12252.0        Not Verified  Dec-2011   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-2011   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-2011   Fully Paid   

  pymnt_plan  \
0          n   
1          n   
2          n   
3          n   
4          n   

                                                                url  \
0  https://lendingclub.com/browse/loanDetail.action?loan_id=1077501   
1  https://lendingclub.com/browse/loanDetail.action?loan_id=1077430   
2  https://lendingclub.com/browse/loanDetail.action?loan_id=1077175   
3  https://lendingclub.com/browse/loanDetail.action?loan_id=1076863   
4  https://lendingclub.com/browse/loanDetail.action?loan_id=1075358   

          purpose                 title zip_code addr_state    dti  \
0     credit_card              Computer    860xx         AZ  27.65   
1             car                  bike    309xx         GA   1.00   
2  small_business  real estate business    606xx         IL   8.72   
3           other              personel    917xx         CA  20.00   
4           other              Personal    972xx         OR  17.94   

   delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
0          0.0         Jan-1985           735.0            739.0   
1          0.0         Apr-1999           740.0            744.0   
2          0.0         Nov-2001           735.0            739.0   
3          0.0         Feb-1996           690.0            694.0   
4          0.0         Jan-1996           695.0            699.0   

   inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc  \
0             1.0                     NaN                     NaN       3.0   
1             5.0                     NaN                     NaN       3.0   
2             2.0                     NaN                     NaN       2.0   
3             1.0                    35.0                     NaN      10.0   
4             0.0                    38.0                     NaN      15.0   

   pub_rec  revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0      0.0    13648.0      83.7%        9.0                   f        0.0   
1      0.0     1687.0       9.4%        4.0                   f        0.0   
2      0.0     2956.0      98.5%       10.0                   f        0.0   
3      0.0     5598.0        21%       37.0                   f        0.0   
4      0.0    27783.0      53.9%       38.0                   f        0.0   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0            0.0   5863.155187          5833.84          5000.00   
1            0.0   1014.530000          1014.53           456.46   
2            0.0   3005.666844          3005.67          2400.00   
3            0.0  12231.89

In [ ]:
# Últimos dados.
dados.tail()

Unnamed: 0         id  loan_amnt  funded_amnt  funded_amnt_inv  \
2925488      105446  102556443    24000.0      24000.0          24000.0   
2925489      105447  102653304    10000.0      10000.0          10000.0   
2925490      105448  102628603    10050.0      10050.0          10050.0   
2925491      105449  102196576     6000.0       6000.0           6000.0   
2925492      105450   99799684    30000.0      30000.0          30000.0   

               term int_rate  installment grade sub_grade  \
2925488   60 months   23.99%       690.30     E        E2   
2925489   36 months    7.99%       313.32     A        A5   
2925490   36 months   16.99%       358.26     D        D1   
2925491   36 months   11.44%       197.69     B        B4   
2925492   60 months   25.49%       889.18     E        E4   

                     emp_title emp_length home_ownership  annual_inc  \
2925488  Senior Escrow Officer   < 1 year           RENT    107000.0   
2925489                     Rn  10+ years       MORTGAGE     65000.0   
2925490        Sales Associate    8 years           RENT     37000.0   
2925491          Contact input    5 years           RENT     41000.0   
2925492      Assistant Manager    4 years       MORTGAGE    105700.0   

        verification_status   issue_d  loan_status pymnt_plan  \
2925488     Source Verified  Apr-2017  Charged Off          n   
2925489     Source Verified  Apr-2017   Fully Paid          n   
2925490        Not Verified  Apr-2017  Charged Off          n   
2925491     Source Verified  Apr-2017   Fully Paid          n   
2925492            Verified  Apr-2017  Charged Off          n   

                                                                        url  \
2925488  https://lendingclub.com/browse/loanDetail.action?loan_id=102556443   
2925489  https://lendingclub.com/browse/loanDetail.action?loan_id=102653304   
2925490  https://lendingclub.com/browse/loanDetail.action?loan_id=102628603   
2925491  https://lendingclub.com/browse/loanDetail.action?loan_id=102196576   
2925492   https://lendingclub.com/browse/loanDetail.action?loan_id=99799684   

                    purpose                    title zip_code addr_state  \
2925488               other                    Other    801xx         CO   
2925489  debt_consolidation       Debt consolidation    187xx         PA   
2925490  debt_consolidation       Debt consolidation    220xx         VA   
2925491         credit_card  Credit card refinancing    105xx         NY   
2925492  debt_consolidation       Debt consolidation    797xx         TX   

           dti  delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
2925488  11.65          3.0         Apr-1995           670.0            674.0   
2925489  19.55          1.0         Sep-1993           725.0            729.0   
2925490  20.56          0.0         May-1993           705.0            709.0   
2925491  19.99          1.0         May-1990           670.0            674.0   
2925492  27.26          1.0         Nov-1997           695.0            699.0   

         inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
2925488             1.0                     4.0                    73.0   
2925489             0.0                    12.0                     NaN   
2925490             1.0                    65.0                     NaN   
2925491             0.0                     0.0                     NaN   
2925492             0.0                     6.0                     NaN   

         open_acc  pub_rec  revol_bal revol_util  total_acc  \
2925488      13.0      2.0     9688.0      24.9%       24.0   
2925489      15.0      0.0     9751.0      15.7%       27.0   
2925490      15.0      0.0    14300.0        47%       21.0   
2925491       9.0      0.0     1356.0      10.1%       18.0   
2925492      12.0      0.0    15252.0      72.6%       23.0   

        initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
2925488                   f        0.0         

In [ ]:
# Verificando a Estatística Básica dos dados.
dados.describe

<bound method NDFrame.describe of          Unnamed: 0         id  loan_amnt  funded_amnt  funded_amnt_inv  \
0                 0    1077501     5000.0       5000.0           4975.0   
1                 1    1077430     2500.0       2500.0           2500.0   
2                 2    1077175     2400.0       2400.0           2400.0   
3                 3    1076863    10000.0      10000.0          10000.0   
4                 4    1075358     3000.0       3000.0           3000.0   
...             ...        ...        ...          ...              ...   
2925488      105446  102556443    24000.0      24000.0          24000.0   
2925489      105447  102653304    10000.0      10000.0          10000.0   
2925490      105448  102628603    10050.0      10050.0          10050.0   
2925491      105449  102196576     6000.0       6000.0           6000.0   
2925492      105450   99799684    30000.0      30000.0          30000.0   

               term int_rate  installment grade sub_grade  \
0   

In [ ]:
# Verificando dados nulos.
dados.isna().sum(axis=0)

Unnamed: 0                                          0
id                                                  0
loan_amnt                                           1
funded_amnt                                         1
funded_amnt_inv                                     1
                                               ...   
hardship_loan_status                          2782082
orig_projected_additional_accrued_interest    2746253
hardship_payoff_balance_amount                2743417
hardship_last_payment_amount                  2743417
debt_settlement_flag                                1
Length: 142, dtype: int64

# 3 - Preparação dos dados
## 3.1 - Selecão dos dados

In [ ]:
# Renomeando as colunas do cicionário de dados.
dicionario = dicionario.rename(columns={'LoanStatNew': 'nome', 'Descrição': 'descrição'})
dicionario.head()

nome  \
0        acc_now_delinq   
1  acc_open_past_24mths   
2            addr_state   
3              all_util   
4            annual_inc   

                                                                 descrição  
0                   O número de contas em que o devedor está inadimplente.  
1                       Número de transações abertas nos últimos 24 meses.  
2                 O estado fornecido pelo mutuário no pedido de empréstimo  
3                     Saldo para limite de crédito em todas as negociações  
4  A renda anual auto-relatada fornecida pelo mutuário durante o registro.

### 3.1.1 - Comparador de dados para seleção de colunas a serem utilizadas

In [ ]:
# Salvando os tipos de dados em um novo DataFrame.
dados_dtypes = pd.DataFrame(dados.dtypes, columns=['dtypes'])
dados_dtypes = dados_dtypes.reset_index()

# Ajuste dos índices e nomes das colunas
dados_dtypes['nome'] = dados_dtypes['index']
dados_dtypes = dados_dtypes[['nome', 'dtypes']]


# Amostra do DataFrame principal para demonstrar como esse dado funciona.
dados_dtypes['primeira_linha'] = dados.loc[0].values
dados_dtypes['linha_meio'] = dados.loc[((len(dados)-1)/2)].values
dados_dtypes['ultima_linha'] = dados.loc[len(dados)-1].values

# Juntando os DataFrames (dtypes e dicionário de dados)
preview = dados_dtypes.merge(dicionario, on='nome', how='left')
preview.head()

nome   dtypes primeira_linha linha_meio ultima_linha  \
0       Unnamed: 0    int64              0     105589       105450   
1               id   object        1077501  160390785     99799684   
2        loan_amnt  float64           5000      17000        30000   
3      funded_amnt  float64           5000      17000        30000   
4  funded_amnt_inv  float64           4975      17000        30000   

                                                                                                                                                             descrição  
0                                                                                                                                                                  NaN  
1                                                                                                        Uma ID exclusiva atribuída ao LC para a lista de empréstimos.  
2  O valor listado do empréstimo solicitado pelo mutuário. Se em algum momento o departamento de crédito reduzir o valor do empréstimo, isso se refletirá neste valor.  
3                                                                                                      O valor total comprometido com esse empréstimo naquele momento.  
4                                                                                O valor total comprometido pelos investidores para aquele empréstimo naquele momento.

In [ ]:
# Conferindo linha por linha para determinar qual coluna permanecer e qual excluir da análise.
preview[135:145]

nome   dtypes primeira_linha  \
135                             hardship_length  float64            NaN   
136                                hardship_dpd  float64            NaN   
137                        hardship_loan_status   object            NaN   
138  orig_projected_additional_accrued_interest  float64            NaN   
139              hardship_payoff_balance_amount  float64            NaN   
140                hardship_last_payment_amount  float64            NaN   
141                        debt_settlement_flag   object              N   

    linha_meio ultima_linha  \
135          4          NaN   
136          0          NaN   
137     ACTIVE          NaN   
138    1300.43          NaN   
139    15456.7          NaN   
140     347.44          NaN   
141          N            N   

                                                                                                                                                                                                                                                descrição  
135                                                                                                                         O número de meses em que o mutuário fará pagamentos menores do que os normalmente obrigados devido a um plano de dificuldades  
136                                                                                                                                                                            Dias da conta vencidos a partir da data de início do plano de dificuldades  
137                                                                                                                                                                                       Status do empréstimo na data de início do plano de dificuldades  
138  A quantia de juros adicional projetada original que será acumulada para o plano de pagamento de dificuldades a partir da Data de início de dificuldades. Este campo será nulo se o mutuário violou o plano de pagamento de dificuldades financeiras.  
139                                                                                                                                                                              O valor do saldo de reembolso na data de início do plano de dificuldades  
140                                                                                                                                                                                O último valor de pagamento na data de início do plano de dificuldades  
141                                                                                                                                               Sinaliza se o mutuário, que efetuou a baixa, está trabalhando com uma empresa de liquidação de dívidas.

In [ ]:
# criando listas com colunas a serem excluidas.
drop_list = ['Unnamed: 0', 'id', 'funded_amnt', 'funded_amnt_inv', 'sub_grade', 'url', 'title', 'zip_code', 'fico_range_low', 
             'fico_range_high', 'last_fico_range_high', 'last_fico_range_low', 'out_prncp_inv', 'total_pymnt_inv', 
             'total_rev_hi_lim', 'tot_hi_cred_lim', 'total_bal_ex_mort', 'total_bc_limit', 'total_il_high_credit_limit', 
             'revol_bal_joint', 'sec_app_fico_range_low', 'sec_app_fico_range_high', 'sec_app_inq_last_6mths', 'sec_app_mort_acc', 
             'sec_app_open_acc', 'sec_app_revol_util', 'sec_app_num_rev_accts', 'sec_app_chargeoff_within_12_mths', 
             'sec_app_collections_12_mths_ex_med', 'hardship_type']

Portanto no momento essas são as colunas que a princípio serão excluídas. Mas pode-se determinar outras colunas a serem excluidas após as análises que serão realizadas.
## 3.2 - Limpeza de dados

In [ ]:
# Verificando informações antes das etapas de limpeza dos dados.
dados.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925493 entries, 0 to 2925492
Columns: 142 entries, Unnamed: 0 to debt_settlement_flag
dtypes: float64(106), int64(1), object(35)
memory usage: 3.1+ GB


In [ ]:
# Verificando a proporção dos dados antes das etapas de limpeza dos dados.
dados.shape

(2925493, 142)

In [ ]:
# Excluindo colunas selecionadas e incluidas na lista a serem dropadas.
dados = dados.drop(drop_list, axis=1)

In [ ]:
# Verificando informações após primeira etapa de limpeza dos dados.
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925493 entries, 0 to 2925492
Columns: 112 entries, loan_amnt to debt_settlement_flag
dtypes: float64(83), object(29)
memory usage: 2.4+ GB


In [ ]:
# Verificando proporção após primeira etapa de limpeza dos dados.
dados.shape

(2925493, 112)

### 3.2.1 - Removendo colunas com muitos dados faltantes

In [ ]:
# Uma maneira inteligente de filtrar dados vazios é usando um limiar de corte (threshold)
meia_amostra = len(dados) / 2
dados = dados.dropna(thresh=meia_amostra, axis=1)

In [ ]:
# Verificando informações após segunda etapa de limpeza dos dados.
dados.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925493 entries, 0 to 2925492
Data columns (total 88 columns):
 #   Column                      Dtype  
---  ------                      -----  
 0   loan_amnt                   float64
 1   term                        object 
 2   int_rate                    object 
 3   installment                 float64
 4   grade                       object 
 5   emp_title                   object 
 6   emp_length                  object 
 7   home_ownership              object 
 8   annual_inc                  float64
 9   verification_status         object 
 10  issue_d                     object 
 11  loan_status                 object 
 12  pymnt_plan                  object 
 13  purpose                     object 
 14  addr_state                  object 
 15  dti                         float64
 16  delinq_2yrs                 float64
 17  earliest_cr_line            object 
 18  inq_last_6mths              float64
 19  open_acc             

In [ ]:
# Verificando proporção após segunda etapa de limpeza dos dados.
dados.shape

(2925493, 88)

In [ ]:
# Salvando os tipos de dados em um novo DataFrame.
dados_dtypes1 = pd.DataFrame(dados.dtypes, columns=['dtypes'])
dados_dtypes1 = dados_dtypes1.reset_index()

# Ajuste dos índices e nomes das colunas
dados_dtypes1['nome'] = dados_dtypes1['index']
dados_dtypes1 = dados_dtypes1[['nome', 'dtypes']]


# Amostra do DataFrame principal para demonstrar como esse dado funciona.
dados_dtypes1['primeira_linha'] = dados.loc[0].values
dados_dtypes1['linha_meio'] = dados.loc[((len(dados)-1)/2)].values
dados_dtypes1['ultima_linha'] = dados.loc[len(dados)-1].values

# Juntando os DataFrames (dtypes e dicionário de dados)
preview1 = dados_dtypes1.merge(dicionario, on='nome', how='left')
preview1.head()

nome   dtypes primeira_linha  linha_meio ultima_linha  \
0    loan_amnt  float64           5000       17000        30000   
1         term   object      36 months   36 months    60 months   
2     int_rate   object         10.65%      16.95%       25.49%   
3  installment  float64         162.87      605.68       889.18   
4        grade   object              B           C            E   

                                                                                                                                                             descrição  
0  O valor listado do empréstimo solicitado pelo mutuário. Se em algum momento o departamento de crédito reduzir o valor do empréstimo, isso se refletirá neste valor.  
1                                                                                O número de pagamentos do empréstimo. Os valores estão em meses e podem ser 36 ou 60.  
2                                                                                                                                          Taxa de juros do empréstimo  
3                                                                                               O pagamento mensal devido pelo mutuário se o empréstimo for originado.  
4                                                                                                                       Nota de empréstimo atribuída pela Lending Club

In [ ]:
# Conferindo linha por linha para determinar qual coluna permanecer e qual excluir da análise.
preview1[73:90]

nome   dtypes primeira_linha linha_meio ultima_linha  \
73             num_il_tl  float64            NaN         14           11   
74         num_op_rev_tl  float64            NaN          6            6   
75         num_rev_accts  float64            NaN         11            9   
76   num_rev_tl_bal_gt_0  float64            NaN          3            4   
77              num_sats  float64            NaN         15           12   
78      num_tl_120dpd_2m  float64            NaN          0            0   
79          num_tl_30dpd  float64            NaN          0            0   
80    num_tl_90g_dpd_24m  float64            NaN          0            0   
81    num_tl_op_past_12m  float64            NaN          5            1   
82        pct_tl_nvr_dlq  float64            NaN        100         91.3   
83      percent_bc_gt_75  float64            NaN         50         66.7   
84  pub_rec_bankruptcies  float64              0          1            0   
85             tax_liens  float64              0          0            0   
86         hardship_flag   object              N          Y            N   
87  debt_settlement_flag   object              N          N            N   

                                                                                                  descrição  
73                                                                              Número de contas parceladas  
74                                                                       Número de contas rotativas abertas  
75                                                                               Número de contas rotativas  
76                                                             Número de negociações rotativas com saldo> 0  
77                                                                           Número de contas satisfatórias  
78                                   Número de contas vencidas há 120 dias (atualizado nos últimos 2 meses)  
79                                    Número de contas vencidas há 30 dias (atualizado nos últimos 2 meses)  
80                                        Número de contas vencidas há 90 ou mais dias nos últimos 24 meses  
81                                                            Número de contas abertas nos últimos 12 meses  
82                                                           Porcentagem de negociações nunca inadimplentes  
83                                        Porcentagem de todas as contas de cartão bancário> 75% do limite.  
84                                                                  Número de falências de registro público  
85                                                                               Número de gravames fiscais  
86                                           Sinaliza se o mutuário está ou não em um plano de dificuldades  
87  Sinaliza se o mutuário, que efetuou a baixa, está trabalhando com uma empresa de liquidação de dívidas.

In [ ]:
# Criando segunda lista com nome de colunas a serem excluídas.
drop_list1 = ['open_acc', 'last_pymnt_d', 'last_pymnt_amnt', 'last_credit_pull_d', 'collections_12_mths_ex_med', 'total_bal_il', 'open_rv_12m', 
              'open_rv_24m', 'max_bal_bc', 'all_util', 'inq_last_12m', 'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'num_bc_sats', 
              'num_rev_accts', 'num_rev_tl_bal_gt_0', 'num_sats', 'num_tl_120dpd_2m', 'percent_bc_gt_75']

In [ ]:
# Excluindo colunas selecionadas e incluidas na lista a serem dropadas.
dados = dados.drop(drop_list1, axis=1)

In [ ]:
# Verificando informações após terceira etapa de limpeza dos dados.
dados.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925493 entries, 0 to 2925492
Data columns (total 68 columns):
 #   Column                    Dtype  
---  ------                    -----  
 0   loan_amnt                 float64
 1   term                      object 
 2   int_rate                  object 
 3   installment               float64
 4   grade                     object 
 5   emp_title                 object 
 6   emp_length                object 
 7   home_ownership            object 
 8   annual_inc                float64
 9   verification_status       object 
 10  issue_d                   object 
 11  loan_status               object 
 12  pymnt_plan                object 
 13  purpose                   object 
 14  addr_state                object 
 15  dti                       float64
 16  delinq_2yrs               float64
 17  earliest_cr_line          object 
 18  inq_last_6mths            float64
 19  pub_rec                   float64
 20  revol_bal               

In [ ]:
# Verificando proporção após terceira etapa de limpeza dos dados.
dados.shape

(2925493, 68)

Pode-se perceber que após a primeira e segunda etapa em que foram excluídas colunas, consideravelmente reduziu-se o número de colunas.

**1. Primeira etapa** -> Colunas selecionadas a serem excluídas.
* Quantidade antes e depois: de 142 para 112.

**2. Segunda etapa** -> Colunas com muitos dados faltantes a serem excluídos.
* Quantidade antes e depois: de 112 para 88.

**3. Terceira etapa** -> Colunas selecionadas a serem excluídas.
* Quantidade antes e depois: de 88 para 68.

Podemos verificar que a maioria de exclusões foram por conta de dados faltantes.

In [ ]:
# Verificando se restaram colunas com dados faltantes.
dados.isna().sum(axis=0)

loan_amnt                   1
term                        1
int_rate                    1
installment                 1
grade                       1
                        ...  
pct_tl_nvr_dlq          70432
pub_rec_bankruptcies     1366
tax_liens                 106
hardship_flag           38436
debt_settlement_flag        1
Length: 68, dtype: int64

In [ ]:
dados.to_csv('lc_dados.csv', index=False)

In [3]:
# Lendo os dados da Leading Club.
df = pd.read_csv('/content/drive/MyDrive/ArquivosPA/limpeza_analise_modelo_Lending_Club/lc_dados.csv', engine='c')

In [4]:
# Informação da base salva.
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925493 entries, 0 to 2925492
Data columns (total 68 columns):
 #   Column                    Dtype  
---  ------                    -----  
 0   loan_amnt                 float64
 1   term                      object 
 2   int_rate                  object 
 3   installment               float64
 4   grade                     object 
 5   emp_title                 object 
 6   emp_length                object 
 7   home_ownership            object 
 8   annual_inc                float64
 9   verification_status       object 
 10  issue_d                   object 
 11  loan_status               object 
 12  pymnt_plan                object 
 13  purpose                   object 
 14  addr_state                object 
 15  dti                       float64
 16  delinq_2yrs               float64
 17  earliest_cr_line          object 
 18  inq_last_6mths            float64
 19  pub_rec                   float64
 20  revol_bal               

## 3.3 - Formatação de dados - Casting (Convertendo tipos de dados em colunas)
Nesta etapa será feita a conversão no tipo de dados em algumas colunas.
* Float -> Int.
* Object -> Datetime.
* Object -> Int

In [5]:
# Verificando os tipos de dados.
print("Tipos de dados e sua frequência\n{}".format(df.dtypes.value_counts()))

Tipos de dados e sua frequência
float64    50
object     18
dtype: int64


### 3.3.1 - Dados de Object para Int, Float e Date

In [6]:
# Mostra as colunas que são do tipo object.
object_columns_df = df.select_dtypes(include=['object'])
print(object_columns_df.iloc[0])

term                      36 months
int_rate                     10.65%
grade                             B
emp_title                       NaN
emp_length                10+ years
home_ownership                 RENT
verification_status        Verified
issue_d                    Dec-2011
loan_status              Fully Paid
pymnt_plan                        n
purpose                 credit_card
addr_state                       AZ
earliest_cr_line           Jan-1985
revol_util                    83.7%
initial_list_status               f
application_type         Individual
hardship_flag                     N
debt_settlement_flag              N
Name: 0, dtype: object


In [7]:
# Alterando formato em que é mostrado as colunas com valores em porcentagens.
df['int_rate'] = df['int_rate'].replace('%', '', regex=True).astype('float') / 100
df['revol_util'] = df['revol_util'].replace('%', '', regex=True).astype('float') / 100

In [8]:
# Verificando os tipos de dados.
print("Tipos de dados e sua frequência\n{}".format(df.dtypes.value_counts()))

Tipos de dados e sua frequência
float64    52
object     16
dtype: int64


In [9]:
# Mostra colunas que estão preenchidas com, no máximo, 4 categorias de dados
for col in df.columns:
    if (len(df[col].unique()) < 4):
        print(df[col].value_counts())
        print()

 36 months    2060077
 60 months     865415
Name: term, dtype: int64

n    2925492
Name: pymnt_plan, dtype: int64

w    2139434
f     786058
Name: initial_list_status, dtype: int64

1.0    2925492
Name: policy_code, dtype: int64

Individual    2714979
Joint App      210513
Name: application_type, dtype: int64

N    2765510
Y     121547
Name: hardship_flag, dtype: int64

N    2874021
Y      51471
Name: debt_settlement_flag, dtype: int64



In [10]:
# Excluir as linhas que a coluna de status do pagamento não sejam "Fully Paid" ou "Charged Off"
loans = df[(df["loan_status"] == "Fully Paid") | (df["loan_status"] == "Charged Off")]

# Gero um dicionário para substituir os status de pagamento
mapping_dictionary = {"loan_status":{ "Fully Paid": 1, "Charged Off": 0}, 
                     "term": {"36 months": 0, "60 months" : 1},
                     "hardship_flag" : {"N":0, "Y":1}}
loans = loans.replace(mapping_dictionary)

In [11]:
# Gero um dicionário para substituir os status de pagamento
mapping_dictionary1 = {"initial_list_status": {"w": 0, "f": 1}, 
                       "debt_settlement_flag": {"N": 0, "Y": 1}}
loans = loans.replace(mapping_dictionary1)

In [12]:
# Mostra colunas que estão preenchidas com mais de 4 categorias de dados
for col in loans.columns:
    if (len(loans[col].unique()) > 4):
        print(loans[col].value_counts())
        print()

10000.0    145211
20000.0    100469
12000.0     98926
15000.0     97754
5000.0      70229
            ...  
700.0           1
950.0           1
38125.0         1
800.0           1
725.0           1
Name: loan_amnt, Length: 1568, dtype: int64

0.1199    42791
0.1099    42601
0.0532    41813
0.1399    38490
0.1149    30621
          ...  
0.1734        1
0.1615        1
0.2264        1
0.1467        1
0.2440        1
Name: int_rate, Length: 684, dtype: int64

301.15     3875
332.10     3417
327.34     3345
361.38     3159
451.73     2749
           ... 
947.41        1
944.59        1
673.21        1
1233.31       1
32.00         1
Name: installment, Length: 90807, dtype: int64

B    546603
C    528107
A    346317
D    275001
E    116500
F     37108
G     10695
Name: grade, dtype: int64

Teacher                           30748
Manager                           27960
Owner                             15903
Registered Nurse                  12592
RN                                12098
   

In [13]:
# Alteração de tipo object e int para float
loans = loans.astype({'int_rate': 'float', 'installment': 'float', 'annual_inc': 'float', 'dti': 'float', 'revol_util': 'float', 
                'total_pymnt': 'float', 'total_rec_prncp': 'float', 'total_rec_int': 'float', 'total_rec_late_fee': 'float', 
                'recoveries': 'float', 'collection_recovery_fee': 'float', 'bc_util': 'float', 
                'pct_tl_nvr_dlq': 'float'})

In [14]:
# Alteração de tipo object, float e int para Int64 
loans['loan_amnt'] = loans['loan_amnt'].astype('float').astype('Int64')
loans['loan_status'] = loans['loan_status'].astype('float').astype('Int64')
loans['delinq_2yrs'] = loans['delinq_2yrs'].astype('float').astype('Int64')
loans['inq_last_6mths'] = loans['inq_last_6mths'].astype('float').astype('Int64')
loans['revol_bal'] = loans['revol_bal'].astype('float').astype('Int64')
loans['total_acc'] = loans['total_acc'].astype('float').astype('Int64')
loans['pub_rec'] = loans['pub_rec'].astype('float').astype('Int64')
loans['acc_now_delinq'] = loans['acc_now_delinq'].astype('float').astype('Int64')
loans['tot_coll_amt'] = loans['tot_coll_amt'].astype('float').astype('Int64')
loans['tot_cur_bal'] = loans['tot_cur_bal'].astype('float').astype('Int64')
loans['open_acc_6m'] = loans['open_acc_6m'].astype('float').astype('Int64')
loans['open_act_il'] = loans['open_act_il'].astype('float').astype('Int64')
loans['open_il_12m'] = loans['open_il_12m'].astype('float').astype('Int64')
loans['open_il_24m'] = loans['open_il_24m'].astype('float').astype('Int64')
loans['mths_since_rcnt_il'] = loans['mths_since_rcnt_il'].astype('float').astype('Int64')
loans['il_util'] = loans['il_util'].astype('float').astype('Int64')
loans['inq_fi'] = loans['inq_fi'].astype('float').astype('Int64')
loans['total_cu_tl'] = loans['total_cu_tl'].astype('float').astype('Int64')
loans['chargeoff_within_12_mths'] = loans['chargeoff_within_12_mths'].astype('float').astype('Int64')
loans['delinq_amnt'] = loans['delinq_amnt'].astype('float').astype('Int64')
loans['mo_sin_old_il_acct'] = loans['mo_sin_old_il_acct'].astype('float').astype('Int64')
loans['mo_sin_old_rev_tl_op'] = loans['mo_sin_old_rev_tl_op'].astype('float').astype('Int64')
loans['mo_sin_rcnt_rev_tl_op'] = loans['mo_sin_rcnt_rev_tl_op'].astype('float').astype('Int64')
loans['mo_sin_rcnt_tl'] = loans['mo_sin_rcnt_tl'].astype('float').astype('Int64')
loans['mths_since_recent_bc'] = loans['mths_since_recent_bc'].astype('float').astype('Int64')
loans['mths_since_recent_inq'] = loans['mths_since_recent_inq'].astype('float').astype('Int64')
loans['num_accts_ever_120_pd'] = loans['num_accts_ever_120_pd'].astype('float').astype('Int64')
loans['num_actv_bc_tl'] = loans['num_actv_bc_tl'].astype('float').astype('Int64')
loans['num_actv_rev_tl'] = loans['num_actv_rev_tl'].astype('float').astype('Int64')
loans['num_bc_tl'] = loans['num_bc_tl'].astype('float').astype('Int64')
loans['num_il_tl'] = loans['num_il_tl'].astype('float').astype('Int64')
loans['num_op_rev_tl'] = loans['num_op_rev_tl'].astype('float').astype('Int64')
loans['num_tl_30dpd'] = loans['num_tl_30dpd'].astype('float').astype('Int64')
loans['num_tl_90g_dpd_24m'] = loans['num_tl_90g_dpd_24m'].astype('float').astype('Int64')
loans['num_tl_op_past_12m'] = loans['num_tl_op_past_12m'].astype('float').astype('Int64')
loans['pub_rec_bankruptcies'] = loans['pub_rec_bankruptcies'].astype('float').astype('Int64')
loans['tax_liens'] = loans['tax_liens'].astype('float').astype('Int64')

In [15]:
# Alteração de tipo object, float e int para Int64
loans['initial_list_status'] = loans['initial_list_status'].astype('Int64')
loans['debt_settlement_flag'] = loans['debt_settlement_flag'].astype('Int64')

In [16]:
# Verificando os tipos de dados.
print("Tipos de dados e sua frequência\n{}".format(loans.dtypes.value_counts()))

Tipos de dados e sua frequência
Int64      39
float64    17
object     12
dtype: int64


In [17]:
# Mostra as colunas que são do tipo object.
object_columns_df = loans.select_dtypes(include=['object'])
print(object_columns_df.iloc[0])

term                     36 months
grade                            B
emp_title                      NaN
emp_length               10+ years
home_ownership                RENT
verification_status       Verified
issue_d                   Dec-2011
pymnt_plan                       n
purpose                credit_card
addr_state                      AZ
earliest_cr_line          Jan-1985
application_type        Individual
Name: 0, dtype: object


In [18]:
cols = ['term', 'grade', 'emp_length', 'home_ownership', 'verification_status', 'addr_state', 'application_type']
for name in cols:
    print(name,':')
    print(object_columns_df[name].value_counts(),'\n')

term :
 36 months    1393299
 60 months     467032
Name: term, dtype: int64 

grade :
B    546603
C    528107
A    346317
D    275001
E    116500
F     37108
G     10695
Name: grade, dtype: int64 

emp_length :
10+ years    611330
2 years      168516
< 1 year     153700
3 years      149271
1 year       122924
5 years      115872
4 years      112211
6 years       85354
8 years       78646
7 years       78141
9 years       67170
Name: emp_length, dtype: int64 

home_ownership :
MORTGAGE    918028
RENT        735358
OWN         205578
ANY           1174
OTHER          144
NONE            49
Name: home_ownership, dtype: int64 

verification_status :
Source Verified    734068
Not Verified       588361
Verified           537902
Name: verification_status, dtype: int64 

addr_state :
CA    264463
TX    152955
NY    150880
FL    133607
IL     72608
NJ     66965
PA     62757
OH     61139
GA     60576
NC     52050
VA     51916
MI     48649
AZ     45508
MD     43578
MA     42709
CO     40492
WA   

In [19]:
# Alterando formato em que é mostrado as colunas como string para Int64.
loans['term'] = loans['term'].replace('months', '', regex=True).replace(' 36', '0', regex=True).replace(' 60', '1', regex=True)
loans['term'] = loans['term'].astype(str).astype(int).astype('Int64')

In [20]:
# Gero um dicionário para substituir os campos emp_length, grade.
mapping_dict = {
    "emp_length": {
        "10+ years": 10,
        "9 years": 9,
        "8 years": 8,
        "7 years": 7,
        "6 years": 6,
        "5 years": 5,
        "4 years": 4,
        "3 years": 3,
        "2 years": 2,
        "1 year": 1,
        "< 1 year": 0,
        "n/a": 0
        },
    "grade":{
        "A": 1,
        "B": 2,
        "C": 3,
        "D": 4,
        "E": 5,
        "F": 6,
        "G": 7
        }
}
loans = loans.replace(mapping_dict)

### 3.3.2 - Balizamento (Salvando dados em bases separadas)

Nesta etapa será realizado separação na base de dados (balizamento), para tornar o dataset (base de dados) mais organizada. Sendo aplicados em colunas que contenham variáveis em que podem ser representadas por códigos, tendo assim um arquivo com extensão `.csv` na qual servirá como dicionário de dados para as colunas que contenham códigos.

In [34]:
# Criando novo DataFrame com valores de iniciais de loan_amnt, home_ownership, addr_state e purpose.
loans_cop = pd.DataFrame(data=loans, columns=['loan_amnt', 'home_ownership', 'addr_state', 'purpose'])
loans_cop.head(2)

loan_amnt home_ownership addr_state      purpose
0       5000           RENT         AZ  credit_card
1       2500           RENT         GA          car

In [35]:
# Criando nova coluna com valores de iniciais de home_ownership, addr_state e purpose.
loans['HomeOwnership_names'] = loans['home_ownership']
loans['addrState_names'] = loans['addr_state']
loans['purpose_names'] = loans['purpose']
loans.head(2)

loan_amnt  term  int_rate  installment  grade emp_title  emp_length  \
0       5000     0    0.1065       162.87      2       NaN        10.0   
1       2500     1    0.1527        59.83      3     Ryder         0.0   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-2011            1   
1           RENT     30000.0     Source Verified  Dec-2011            0   

  pymnt_plan      purpose addr_state    dti  delinq_2yrs earliest_cr_line  \
0          n  credit_card         AZ  27.65            0         Jan-1985   
1          n          car         GA   1.00            0         Apr-1999   

   inq_last_6mths  pub_rec  revol_bal  revol_util  total_acc  \
0               1        0      13648       0.837          9   
1               5        0       1687       0.094          4   

   initial_list_status  out_prncp  total_pymnt  total_rec_prncp  \
0                    1        0.0  5863.155187          5000.00   
1                    1        0.0  1014.530000           456.46   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         863.16                 0.0         0.0                     0.00   
1         435.17                 0.0       122.9                     1.11   

   policy_code application_type  acc_now_delinq  tot_coll_amt  tot_cur_bal  \
0          1.0       Individual               0          <NA>         <NA>   
1          1.0       Individual               0          <NA>         <NA>   

   open_acc_6m  open_act_il  open_il_12m  open_il_24m  mths_since_rcnt_il  \
0         <NA>         <NA>         <NA>         <NA>                <NA>   
1         <NA>         <NA>         <NA>         <NA>                <NA>   

   il_util  inq_fi  total_cu_tl  bc_util  chargeoff_within_12_mths  \
0     <NA>    <NA>         <NA>      NaN                         0   
1     <NA>    <NA>         <NA>      NaN                         0   

   delinq_amnt  mo_sin_old_il_acct  mo_sin_old_rev_tl_op  \
0            0                <NA>                  <NA>   
1            0                <NA>                  <NA>   

   mo_sin_rcnt_rev_tl_op  mo_sin_rcnt_tl  mort_acc  mths_since_recent_bc  \
0                   <NA>            <NA>       NaN                  <NA>   
1                   <NA>            <NA>       NaN                  <NA>   

   mths_since_recent_inq  num_accts_ever_120_pd  num_actv_bc_tl  \
0                   <NA>                   <NA>            <NA>   
1                   <NA>                   <NA>            <NA>   

   num_actv_rev_tl  num_bc_tl  num_il_tl  num_op_rev_tl  num_tl_30dpd  \
0             <NA>       <NA>       <NA>           <NA>          <NA>   
1             <NA>       <NA>       <NA>           <NA>          <NA>   

   num_tl_90g_dpd_24m  num_tl_op_past_12m  pct_tl_nvr_dlq  \
0                <NA>                <NA>             NaN   
1                <NA>                <NA>             NaN   

   pub_rec_bankruptcies  tax_liens  hardship_flag  debt_settlement_flag  \
0                     0          0            0.0                     0   
1                     0          0            0.0                     0   

  purpose_names HomeOwnership_names addrState_names  
0   credit_card                RENT              AZ  
1           car                RENT              GA

In [36]:
# Gero um dicionário para substituir os campos de home_ownership, addr_state e purpose.
mapping_dict1 = {
    "home_ownership": {
        "RENT": 1,
        "OWN": 2,
        "MORTGAGE": 3, 
        "OTHER": 4, 
        "NONE": 5, 
        "ANY": 6
    },
    "addr_state": {
        "AZ": 1, "GA": 2, "IL": 3, "CA": 4, "OR": 5, "NC": 6, "TX": 7, "VA": 8, "MO": 9, "CT": 10, 
        "UT": 11, "FL": 12, "NY": 13, "PA": 14, "MN": 15, "NJ": 16, "KY": 17, "OH": 18, "SC": 19, 
        "RI": 20, "LA": 21, "MA": 22, "WA": 23, "WI": 24, "AL": 25, "CO": 26, "KS": 27, "NV": 28, 
        "AK": 29, "MD": 30, "WV": 31, "VT": 32, "MI": 33, "DC": 34, "SD": 35, "NH": 36, "AR": 37, 
        "NM": 38, "MT": 39, "HI": 40, "WY": 41, "OK": 42, "DE": 43, "MS": 44, "TN": 45, "IA": 46, 
        "NE": 47, "ID": 48, "IN": 49, "ME": 50, "ND": 51
    },
    "purpose": {
        "credit_card": 1, "car": 2, 
        "small_business": 3, "other": 4, 
        "wedding": 5, "debt_consolidation": 6, 
        "home_improvement": 7, "major_purchase": 8, 
        "medical": 9, "moving": 10, "vacation": 11, 
        "house": 12, "renewable_energy": 13, 
        "educational": 14
    }
}
loans = loans.replace(mapping_dict1)

In [37]:
# Verificando se foi realizada a alteração corretamente.
loans.head(2)

loan_amnt  term  int_rate  installment  grade emp_title  emp_length  \
0       5000     0    0.1065       162.87      2       NaN        10.0   
1       2500     1    0.1527        59.83      3     Ryder         0.0   

   home_ownership  annual_inc verification_status   issue_d  loan_status  \
0               1     24000.0            Verified  Dec-2011            1   
1               1     30000.0     Source Verified  Dec-2011            0   

  pymnt_plan  purpose  addr_state    dti  delinq_2yrs earliest_cr_line  \
0          n        1           1  27.65            0         Jan-1985   
1          n        2           2   1.00            0         Apr-1999   

   inq_last_6mths  pub_rec  revol_bal  revol_util  total_acc  \
0               1        0      13648       0.837          9   
1               5        0       1687       0.094          4   

   initial_list_status  out_prncp  total_pymnt  total_rec_prncp  \
0                    1        0.0  5863.155187          5000.00   
1                    1        0.0  1014.530000           456.46   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         863.16                 0.0         0.0                     0.00   
1         435.17                 0.0       122.9                     1.11   

   policy_code application_type  acc_now_delinq  tot_coll_amt  tot_cur_bal  \
0          1.0       Individual               0          <NA>         <NA>   
1          1.0       Individual               0          <NA>         <NA>   

   open_acc_6m  open_act_il  open_il_12m  open_il_24m  mths_since_rcnt_il  \
0         <NA>         <NA>         <NA>         <NA>                <NA>   
1         <NA>         <NA>         <NA>         <NA>                <NA>   

   il_util  inq_fi  total_cu_tl  bc_util  chargeoff_within_12_mths  \
0     <NA>    <NA>         <NA>      NaN                         0   
1     <NA>    <NA>         <NA>      NaN                         0   

   delinq_amnt  mo_sin_old_il_acct  mo_sin_old_rev_tl_op  \
0            0                <NA>                  <NA>   
1            0                <NA>                  <NA>   

   mo_sin_rcnt_rev_tl_op  mo_sin_rcnt_tl  mort_acc  mths_since_recent_bc  \
0                   <NA>            <NA>       NaN                  <NA>   
1                   <NA>            <NA>       NaN                  <NA>   

   mths_since_recent_inq  num_accts_ever_120_pd  num_actv_bc_tl  \
0                   <NA>                   <NA>            <NA>   
1                   <NA>                   <NA>            <NA>   

   num_actv_rev_tl  num_bc_tl  num_il_tl  num_op_rev_tl  num_tl_30dpd  \
0             <NA>       <NA>       <NA>           <NA>          <NA>   
1             <NA>       <NA>       <NA>           <NA>          <NA>   

   num_tl_90g_dpd_24m  num_tl_op_past_12m  pct_tl_nvr_dlq  \
0                <NA>                <NA>             NaN   
1                <NA>                <NA>             NaN   

   pub_rec_bankruptcies  tax_liens  hardship_flag  debt_settlement_flag  \
0                     0          0            0.0                     0   
1                     0          0            0.0                     0   

  purpose_names HomeOwnership_names addrState_names  
0   credit_card                RENT              AZ  
1           car                RENT              GA

In [ ]:
#  Separando coluna home_ownership e criando dicionário desta.
homeOwnership = loans[['HomeOwnership_names', 'home_ownership']]
homeOwnership = homeOwnership.groupby(by=['HomeOwnership_names', 'home_ownership']).count().reset_index()
homeOwnership = homeOwnership.rename(columns={'HomeOwnership_names': 'grupo', 'home_ownership': 'codigo'})

# Salvando em arquivo .csv
homeOwnership.to_csv('home_ownership.csv', index=False)

In [ ]:
#  Separando coluna addr_state e criando dicionário desta.
addrState = loans[['addrState_names', 'addr_state']]
addrState = addrState.groupby(by=['addrState_names', 'addr_state']).count().reset_index()
addrState = addrState.rename(columns={'addrState_names': 'grupo', 'addr_state': 'codigo'})

# Salvando em arquivo .csv
addrState.to_csv('addr_state.csv', index=False)

In [45]:
#  Separando coluna purpose e criando dicionário desta.
purpose_grupo = loans[['purpose_names', 'purpose']]
purpose_grupo = purpose_grupo.groupby(by=['purpose_names', 'purpose']).count().reset_index()
purpose_grupo = purpose_grupo.rename(columns={'purpose_names': 'grupo', 'purpose': 'codigo'})

# Salvando em arquivo .csv
purpose_grupo.to_csv('purpose.csv', index=False)

In [48]:
# Criando lista das colunas alterada a serem excluídas.
drop_lista = ['purpose_names', 'HomeOwnership_names', 'addrState_names']

# Excluindo colunas selecionadas e incluidas na lista a serem dropadas.
loans = loans.drop(drop_lista, axis=1)

# Verificando se foram excluídas corretamente.
loans.head(2)

loan_amnt  term  int_rate  installment  grade emp_title  emp_length  \
0       5000     0    0.1065       162.87      2       NaN        10.0   
1       2500     1    0.1527        59.83      3     Ryder         0.0   

   home_ownership  annual_inc verification_status   issue_d  loan_status  \
0               1     24000.0            Verified  Dec-2011            1   
1               1     30000.0     Source Verified  Dec-2011            0   

  pymnt_plan  purpose  addr_state    dti  delinq_2yrs earliest_cr_line  \
0          n        1           1  27.65            0         Jan-1985   
1          n        2           2   1.00            0         Apr-1999   

   inq_last_6mths  pub_rec  revol_bal  revol_util  total_acc  \
0               1        0      13648       0.837          9   
1               5        0       1687       0.094          4   

   initial_list_status  out_prncp  total_pymnt  total_rec_prncp  \
0                    1        0.0  5863.155187          5000.00   
1                    1        0.0  1014.530000           456.46   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         863.16                 0.0         0.0                     0.00   
1         435.17                 0.0       122.9                     1.11   

   policy_code application_type  acc_now_delinq  tot_coll_amt  tot_cur_bal  \
0          1.0       Individual               0          <NA>         <NA>   
1          1.0       Individual               0          <NA>         <NA>   

   open_acc_6m  open_act_il  open_il_12m  open_il_24m  mths_since_rcnt_il  \
0         <NA>         <NA>         <NA>         <NA>                <NA>   
1         <NA>         <NA>         <NA>         <NA>                <NA>   

   il_util  inq_fi  total_cu_tl  bc_util  chargeoff_within_12_mths  \
0     <NA>    <NA>         <NA>      NaN                         0   
1     <NA>    <NA>         <NA>      NaN                         0   

   delinq_amnt  mo_sin_old_il_acct  mo_sin_old_rev_tl_op  \
0            0                <NA>                  <NA>   
1            0                <NA>                  <NA>   

   mo_sin_rcnt_rev_tl_op  mo_sin_rcnt_tl  mort_acc  mths_since_recent_bc  \
0                   <NA>            <NA>       NaN                  <NA>   
1                   <NA>            <NA>       NaN                  <NA>   

   mths_since_recent_inq  num_accts_ever_120_pd  num_actv_bc_tl  \
0                   <NA>                   <NA>            <NA>   
1                   <NA>                   <NA>            <NA>   

   num_actv_rev_tl  num_bc_tl  num_il_tl  num_op_rev_tl  num_tl_30dpd  \
0             <NA>       <NA>       <NA>           <NA>          <NA>   
1             <NA>       <NA>       <NA>           <NA>          <NA>   

   num_tl_90g_dpd_24m  num_tl_op_past_12m  pct_tl_nvr_dlq  \
0                <NA>                <NA>             NaN   
1                <NA>                <NA>             NaN   

   pub_rec_bankruptcies  tax_liens  hardship_flag  debt_settlement_flag  
0                     0          0            0.0                     0  
1                     0          0            0.0                     0

In [49]:
# Detecta colunas preenchidas com um único valor
loans = loans.loc[:,loans.apply(pd.Series.nunique) != 1]

### 3.3.3 - Gerando dummies

In [50]:
# Gero dummies automaticamente
nominal_columns = ["verification_status", "application_type"]
dummy_df = pd.get_dummies(loans[nominal_columns])
loans = pd.concat([loans, dummy_df], axis=1)
loans = loans.drop(nominal_columns, axis=1)

In [52]:
# Verificando os tipos de dados.
print("Tipos de dados e sua frequência\n{}".format(loans.dtypes.value_counts()))

Tipos de dados e sua frequência
Int64      40
float64    16
uint8       5
int64       4
object      3
dtype: int64


In [53]:
# Mostra as colunas que são do tipo object.
object_columns_df = loans.select_dtypes(include=['object'])
print(object_columns_df.iloc[0])

emp_title                NaN
issue_d             Dec-2011
earliest_cr_line    Jan-1985
Name: 0, dtype: object


In [59]:
# Alterando formato em que é mostrado nas colunas como mês nome para mês numérico.
loans['issue_d'] = loans['issue_d'].replace('Jan', '01', regex=True).replace('Feb', '02', regex=True).replace('Mar', '03', regex=True).replace('Apr', '04', regex=True).replace('May', '05', regex=True).replace('Jun', '06', regex=True).replace('Jul', '07', regex=True).replace('Aug', '08', regex=True).replace('Sep', '09', regex=True).replace('Oct', '10', regex=True).replace('Nov', '11', regex=True).replace('Dec', '12', regex=True)

loans['earliest_cr_line'] = loans['earliest_cr_line'].replace('Jan', '01', regex=True).replace('Feb', '02', regex=True).replace('Mar', '03', regex=True).replace('Apr', '04', regex=True).replace('May', '05', regex=True).replace('Jun', '06', regex=True).replace('Jul', '07', regex=True).replace('Aug', '08', regex=True).replace('Sep', '09', regex=True).replace('Oct', '10', regex=True).replace('Nov', '11', regex=True).replace('Dec', '12', regex=True)

In [65]:
# Convertendo para datetime.
loans['issue_d'] = pd.to_datetime(loans['issue_d'], format='%m-%Y')
loans['earliest_cr_line'] = pd.to_datetime(loans['earliest_cr_line'], format='%m-%Y')

In [70]:
# Mostra as colunas que são do tipo object.
object_columns_df = loans.select_dtypes(include=['object'])
print(object_columns_df.iloc[0])

emp_title    NaN
Name: 0, dtype: object


In [73]:
# Verificando se restaram colunas com dados faltantes ou nulos, mostrando por slice.
loans.isna().sum(axis=0)[:20]

loan_amnt                0
term                     0
int_rate                 0
installment              0
grade                    0
emp_title           131454
emp_length          117196
home_ownership           0
annual_inc               0
issue_d                  0
loan_status              0
purpose                  0
addr_state               0
dti                   1108
delinq_2yrs              0
earliest_cr_line         0
inq_last_6mths           1
pub_rec                  0
revol_bal                0
revol_util            1398
dtype: int64

In [ ]:
# Alterando valores nulos ou faltantes para 0
loans['percent_bc_gt_75'] = loans['percent_bc_gt_75'].fillna(0.0)
loans['pub_rec_bankruptcies'] = loans['pub_rec_bankruptcies'].fillna(0)
loans['tax_liens'] = loans['tax_liens'].fillna(0)
loans['hardship_flag'] = loans['hardship_flag'].fillna(0)
loans['annual_inc'] = loans['annual_inc'].fillna(0)

In [ ]:
# Alteração de tipo object, float e int para Int64 
loans['percent_bc_gt_75'] = loans['percent_bc_gt_75'].astype('float')
loans['pub_rec_bankruptcies'] = loans['pub_rec_bankruptcies'].astype('float').astype(int).astype('Int64')
loans['tax_liens'] = loans['tax_liens'].astype('float').astype(int).astype('Int64')
loans['hardship_flag'] = loans['hardship_flag'].astype('float').astype(int).astype('Int64')
loans['annual_inc'] = loans['annual_inc'].astype('float').astype(int).astype('Int64')

OBS: Não sei se está é a melhor opção para tirar os dados nulos ou faltantes restantes. Portanto no momento irá permanecer como está, até uma verificação mais precisa de como proceder.

# 4 - O resultado da limpeza deve ser salvo no formato .pikle.

In [74]:
# Salvando em arquivo .pkl
loans.to_pickle('dados_lc_limpos.pkl') 